Src/Plugins/Security/SEC101_020.DropboxAccessTokenValidator.cs

```diff
@@ -0,0 +1,133 @@
+        {
+            if (!groups.TryGetNonEmptyValue("refine", out string key))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            if (!ContainsDigitAndChar(key))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            fingerprintText = new Fingerprint()
+            {
+                Key = key,
+                Platform = nameof(AssetPlatform.Dropbox),
+            }.ToString();
+
+            return nameof(ValidationState.Unknown);
+        }
+
+        protected override string IsValidDynamicHelper(ref string fingerprintText, ref string message, ref Dictionary<string, string> options)
+        {
+            var fingerprint = new Fingerprint(fingerprintText);
+            string key = fingerprint.Key;
+            using HttpClient httpClient = CreateHttpClient();
+            httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", key);
+
+            try
+            {
+                using HttpResponseMessage response = httpClient
+                    .PostAsync("https://api.dropboxapi.com/2/file_requests/count", null)
+                    .GetAwaiter()
+                    .GetResult();
+
+                switch (response.StatusCode)
+                {
+                    case HttpStatusCode.OK:
+                    {
+                        return nameof(ValidationState.AuthorizedError);
+                    }
+
+                    case HttpStatusCode.BadRequest:
+                    {
+                        string body = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
+
+                        // App deleted.
+                        if (body.Contains("This app is currently disabled"))
+                        {
+                            return nameof(ValidationState.Expired);
+                        }
+
+                        // Request was successfull but AccessToken does not have access.
+                        if (body.Contains("file_requests.read"))
+                        {
+                            return nameof(ValidationState.AuthorizedError);
+                        }
+
+                        // We don't recognize this message.
+                        message = CreateUnexpectedResponseCodeMessage(response.StatusCode);
+                        break;
+                    }
+
+                    case HttpStatusCode.Unauthorized:
+                    {
+                        return nameof(ValidationState.Unauthorized);
+                    }
+
+                    default:
+                    {
+                        message = CreateUnexpectedResponseCodeMessage(response.StatusCode);
+                        break;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                return ReturnUnhandledException(ref message, e);
+            }
+
+            return nameof(ValidationState.Unknown);
```

Comment:
>nameof [](start = 19, length = 6)

If you move this to the default case, I think this code path becomes unfeasible and you wouldn't be able to have anything here. this would co-locate this return path with the message construction. probably a clearer pattern to use (I know, of course, that we have preferred the pattern you have here in other rules. just a thought to explore, not a necessary change.

Reply:
I didn't do that because there are two places that i use this:
1. default case
2. bad request not specified message.  so, i would leave this for now and think in a  better strategy later :)

---
In reply to: [598008839](https://github.com/microsoft/sarif-pattern-matcher/pull/315#discussion_r598008839) [](ancestors = 598008839)